Fix: Add context support to Consolidator. Wait to prevent goroutine leaks (#18272)

AshishSatish opened the pull request:
## Description

This pull request updates the `PendingResult.Wait()` method in the query consolidator to support `context.Context`, allowing waiting goroutines to respect cancellation or timeout.

Previously, if the original query being consolidated did not return (e.g. due to a hang or mismanaged timeout), duplicate queries would block indefinitely on `sync.RWMutex.RLock()`. This led to goroutine buildup, memory pressure, and excessive GC cycles, as reported in [Issue #18272](https://github.com/vitessio/vitess/issues/18272).

### What this PR changes:
- Updated the `PendingResult` interface in `consolidator.go` to change `Wait()` → `Wait(ctx context.Context) error`
- Wrapped the `.RLock()` call in a goroutine, and used a `select` to allow returning early if the context is done
- Safely releases the lock if the context is cancelled before `.RLock()` is acquired
- Updated the `FakePendingResult` mock in `fake_consolidator.go` to match the new method signature

This is a safe and minimal fix that improves consolidator robustness and prevents goroutine leaks in edge cases.

## Related Issue(s)

- Fixes [#18272](https://github.com/vitessio/vitess/issues/18272)

## Checklist

- [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
- [x] If this change is to be back-ported to previous releases, a justification is included in the PR description *(not required in this case)*
- [x] Tests were added or are not required *(interface is not currently called externally; mock updated)*
- [x] Did the new or modified tests pass consistently locally and on CI?
- [x] Documentation was added or is not required *(this is an internal concurrency fix)*

## Deployment Notes

No deployment action required. This change affects internal consolidator logic and interface only.

dbussink commented on go/sync2/consolidator.go:
```diff
@@ -117,13 +117,31 @@ func (rs *pendingResult) SetResult(res *sqltypes.Result) {
-	rs.executing.RLock()
+	acquired := make(chan struct{})
+	go func() {
+		rs.executing.RLock()
```
How does this stop the this new goroutine from leaking? We'd now wait indefinitely on this new lock statement here? 

Have you validated that this solves the problem and that you don't see the lake? If so, how? We also need a test for this to validate it works.

dbussink commented on go/sync2/consolidator.go:
```diff
@@ -117,13 +117,31 @@ func (rs *pendingResult) SetResult(res *sqltypes.Result) {
+	}
 }
 
+
```
Don't add stray newlines like this.

AshishSatish commented:
Thank you for the feedback @dbussink — you're absolutely right.

In the earlier version, the goroutine always called `RLock()` even if the context was already canceled, which could still result in a leak.

I've now updated the implementation to:
- Check `ctx.Done()` before calling `RLock()`, preventing unnecessary blocking
- Avoid any goroutine leak if the context is already expired
- Also cleaned up the extra newline you mentioned
- Added a test (`TestPendingResult_Wait_ContextTimeout`) to validate that the goroutine exits properly on context timeout

Let me know if any further changes are needed — happy to adjust!

AshishSatish commented:
Thanks for the honest feedback. I realize I need to get a better handle on concurrency and locking in Go before I can tackle this properly. Going to step back from this PR for now, but I really appreciate the guidance — hope to come back stronger next time!
